### PR TITLE
Activity Aggregate refactoring

### PR DESCRIPTION
**Pull request by maheshkumargangula**

### CODE REVIEW

File: activity-aggregate-updater/src/main/scala/org/sunbird/job/functions/ActivityAggregatesFunction.scala [hunk 1]
```diff
@@ -97,24 +102,57 @@ class ActivityAggregatesFunction(config: ActivityAggregateUpdaterConfig)(implici
...
+    val contents = event.getOrElse(config.contents, new util.ArrayList[java.util.Map[String, AnyRef]]()).asInstanceOf[util.List[java.util.Map[String, AnyRef]]].asScala
+    if (contents.size > 0) {
+      val content =  contents.head
+      val contentId = content.getOrDefault("contentId", "").asInstanceOf[String]
+      val status = content.getOrDefault("status", 0.asInstanceOf[AnyRef]).asInstanceOf[Number].intValue()
+      val checksum = getMessageId(courseId, batchId, userId, contentId, status)
+      val isUnique = deDupEngine.isUniqueEvent(checksum)
+      if (isUnique) deDupEngine.storeChecksum(checksum)
+      isUnique
+    } else false
```
Comment (SanthoshVasabhaktula) [inline activity-aggregate-updater/src/main/scala/org/sunbird/job/functions/ActivityAggregatesFunction.scala:144]: unique check should be done for all contents, not the first one. And there may be possibility that some of the contents within an event can be duplicate. The method needs to change to discardDuplicates(event) which would return a List of userId, courseId, batchId, contentId, status. This should also filter out the events with status=1

**Review by SanthoshVasabhaktula — CHANGES_REQUESTED** [review]

File: activity-aggregate-updater/src/main/scala/org/sunbird/job/functions/ActivityAggregatesFunction.scala [hunk 2]
```diff
@@ -15,67 +16,84 @@ import org.apache.flink.streaming.api.functions.windowing.ProcessWindowFunction
...
+      if (!isBatchEnrollmentEvent) metrics.incCounter(config.skipEventsCount)
       isBatchEnrollmentEvent
+    }.flatMap { event =>
+      val contents = event.getOrElse(config.contents, new util.ArrayList[java.util.Map[String, AnyRef]]()).asInstanceOf[util.List[java.util.Map[String, AnyRef]]].asScala
+      val filteredContents = contents.filter(x => x.get("status") == 2).toList
+      if (filteredContents.size == 0) metrics.incCounter(config.skipEventsCount)
+      filteredContents.map(c => {
+        event + ("contents" -> List(Map("contentId" -> c.get("contentId"), "status" -> c.get("status"))))
+      })
     }
```
Comment (SanthoshVasabhaktula) [inline activity-aggregate-updater/src/main/scala/org/sunbird/job/functions/ActivityAggregatesFunction.scala:77]: Convert this into a case class of (batchId, courseId, contentId, userId, status)

File: activity-aggregate-updater/src/main/scala/org/sunbird/job/task/ActivityAggregateUpdaterStreamTask.scala [hunk 3]
```diff
@@ -21,20 +23,30 @@ class ActivityAggregateUpdaterStreamTask(config: ActivityAggregateUpdaterConfig,
...
     implicit val stringTypeInfo: TypeInformation[String] = TypeExtractor.getForClass(classOf[String])
+    implicit val enrolmentCompleteTypeInfo: TypeInformation[List[EnrolmentComplete]] = TypeExtractor.getForClass(classOf[List[EnrolmentComplete]])
     val progressStream =
       env.addSource(kafkaConnector.kafkaMapSource(config.kafkaInputTopic), config.activityAggregateUpdaterConsumer)
         .uid(config.activityAggregateUpdaterConsumer).setParallelism(config.kafkaConsumerParallelism)
         .rebalance()
-        .keyBy(x => x.get("partition").toString)
+        // .keyBy(x => x.get("partition").toString)
+        .keyBy(new ActivityAggregatorKeySelector(config))
         .timeWindow(Time.seconds(config.thresholdBatchReadInterval))
```
Comment (SanthoshVasabhaktula) [inline activity-aggregate-updater/src/main/scala/org/sunbird/job/task/ActivityAggregateUpdaterStreamTask.scala:33]: Why don't we do a key by userId modulus 2000 (shards) and do a countWindow(100)? @maheshkumargangula @anandp504

**Review by SanthoshVasabhaktula — CHANGES_REQUESTED** [review]

File: activity-aggregate-updater/src/main/scala/org/sunbird/job/task/ActivityAggregateUpdaterStreamTask.scala (hunk 3 above)
Comment (anandp504) [reply activity-aggregate-updater/src/main/scala/org/sunbird/job/task/ActivityAggregateUpdaterStreamTask.scala:33]: @SanthoshVasabhaktula I am working on making the change to use a mod of userId as the partition key.

**Pull request merged**